File: leetcode/design_in_memory_file_system/design_in_memory_file_system.cpp
/*
 * Time Complexity:
 *   - ls, mkdir, addContentToFile, readContentFromFile: O(L + K log K)
 *     - L is the number of parts in the path (depth of the file/directory).
 *     - K is the number of files/directories in the current directory (for ls sorting).
 *     - Each path operation traverses at most L nodes.
 *
 * Space Complexity: O(N)
 *   - N is the total number of files and directories created.
 *   - Space is used for the tree structure and file contents.
 */

#include <unordered_map>
#include <vector>
#include <string>
#include <sstream>
#include <algorithm>

using std::string;
using std::vector;
using std::unordered_map;
using std::stringstream;
using std::sort;

/**
 * Your FileSystem object will be instantiated and called as such:
 * FileSystem obj = new FileSystem();
 * vector<string> param_1 = obj.ls(path);
 * obj.mkdir(path);
 * obj.addContentToFile(filePath,content);
 * string param_4 = obj.readContentFromFile(filePath);
 */
struct FileNode {
    bool isFile;
    string content;
    unordered_map<string, FileNode*> children;
    FileNode() : isFile(false) {}
};

class FileSystem {
private:
    FileNode* root;

    vector<string> split(const string& s, const char delim) {
        vector<string> tokens;
        stringstream ss(s);
        string token;
        while (getline(ss, token, delim)) {
            if (!token.empty()) {
                tokens.emplace_back(token);
            }
        }
        return tokens;
    }

    FileNode* getFileNode(const string& path) {
        FileNode* currNode = root;
        vector<string> parts = split(path, '/');
        for (const string& part : parts) {
            currNode = currNode->children[part];
        }
        return currNode;
    }

    FileNode* putFileNode(const string& path) {
        FileNode* currNode = root;
        vector<string> parts = split(path, '/');
        for (const string& part : parts) {
            if (currNode->children.find(part) == currNode->children.end()) {
                currNode->children[part] = new FileNode();
            }
            currNode = currNode->children[part];
        }
        return currNode;
    }

public:
    FileSystem() {
        root = new FileNode();
    }

    vector<string> ls(string path) {
        FileNode* currNode = getFileNode(path);
        if(currNode->isFile) {
            return {split(path, '/').back()};
        }

        vector<string> ans;
        for(const auto& child : currNode->children) {
            ans.emplace_back(child.first);
        }
        sort(ans.begin(), ans.end());

        return ans;
    }

    void mkdir(string path) {
        putFileNode(path);
    }

    void addContentToFile(string filePath, string content) {
        FileNode* fileNode = putFileNode(filePath);
        fileNode->isFile = true;
        fileNode->content += content;
    }

    string readContentFromFile(string filePath) {
        return getFileNode(filePath)->content;
    }
};

File: leetcode/design_in_memory_file_system/design_in_memory_file_system.java
package design_in_memory_file_system;

import java.util.*;

class FileNode {
    boolean isFile = false;
    StringBuilder content = new StringBuilder();
    Map<String, FileNode> children = new HashMap<>();
}

class FileSystem {
    private final FileNode root;

    public FileSystem() {
        root = new FileNode();
    }

    private FileNode getFileNode(String path) {
        FileNode curr = root;
        if ("/".equals(path)) return curr;
        String[] parts = path.split("/");
        for (String part : parts) {
            if (part.isEmpty()) continue;
            curr = curr.children.get(part);
        }
        return curr;
    }

    private FileNode putFileNode(String path) {
        FileNode curr = root;
        String[] parts = path.split("/");
        for (String part : parts) {
            if (part.isEmpty()) continue;
            curr.children.putIfAbsent(part, new FileNode());
            curr = curr.children.get(part);
        }
        return curr;
    }

    public List<String> ls(String path) {
        FileNode node = getFileNode(path);
        if (node.isFile) {
            String[] parts = path.split("/");
            return Arrays.asList(parts[parts.length - 1]);
        }
        List<String> res = new ArrayList<>(node.children.keySet());
        Collections.sort(res);
        return res;
    }

    public void mkdir(String path) {
        putFileNode(path);
    }

    public void addContentToFile(String filePath, String content) {
        FileNode node = putFileNode(filePath);
        node.isFile = true;
        node.content.append(content);
    }

    public String readContentFromFile(String filePath) {
        return getFileNode(filePath).content.toString();
    }
}